builtin/logical/database/credentials.go

```diff
@@ -170,3 +175,190 @@ func (kg rsaKeyGenerator) configMap() (map[string]interface{}, error) {
+		return cg, err
+	}
+
+	switch cg.KeyType {
+	case "rsa":
+		switch cg.KeyBits {
+		case 0, 2048, 3072, 4096:
+		default:
+			return cg, fmt.Errorf("invalid key_bits")
+		}
+	case "ec":
+		switch cg.KeyBits {
+		case 0, 224, 256, 384, 521:
+		default:
+			return cg, fmt.Errorf("invalid key_bits")
+		}
+	case "ed25519":
+		// key_bits ignored
+	default:
+		return cg, fmt.Errorf("invalid key_type")
+	}
+
+	switch cg.SignatureBits {
+	case 0, 256, 384, 512:
+	default:
+		return cg, fmt.Errorf("invalid signature_bits")
+	}
+
+	if cg.CommonNameTemplate == "" {
+		return cg, fmt.Errorf("missing required common_name_template")
+	}
+
+	// Validate the common name template
+	t, err := template.NewTemplate(template.Template(cg.CommonNameTemplate))
+	if err != nil {
+		return cg, fmt.Errorf("failed to create template: %w", err)
+	}
+
+	_, err = t.Generate(dbplugin.UsernameMetadata{})
+	if err != nil {
+		return cg, fmt.Errorf("invalid common_name_template: %w", err)
+	}
+	cg.cnProducer = t
+
+	if cg.CACert == "" {
+		return cg, fmt.Errorf("missing required ca_cert")
+	}
+	if cg.CAPrivateKey == "" {
+		return cg, fmt.Errorf("missing required ca_private_key")
+	}
+	parsedBundle, err := certutil.ParsePEMBundle(strings.Join([]string{cg.CACert, cg.CAPrivateKey}, "\n"))
+	if err != nil {
+		return cg, err
+	}
+	if parsedBundle.PrivateKey == nil {
+		return cg, fmt.Errorf("private key not found in the PEM bundle")
+	}
+	if parsedBundle.PrivateKeyType == certutil.UnknownPrivateKey {
+		return cg, fmt.Errorf("unknown private key found in the PEM bundle")
+	}
+	if parsedBundle.Certificate == nil {
+		return cg, fmt.Errorf("certificate not found in the PEM bundle")
+	}
+	if !parsedBundle.Certificate.IsCA {
+		return cg, fmt.Errorf("the given certificate is not marked for CA use")
+	}
+
+	certBundle, err := parsedBundle.ToCertBundle()
+	if err != nil {
+		return cg, fmt.Errorf("error converting raw values into cert bundle: %w", err)
+	}
+
+	parsedCABundle, err := certBundle.ToParsedCertBundle()
+	if err != nil {
+		return cg, fmt.Errorf("failed to parse cert bundle: %w", err)
+	}
+	cg.parsedCABundle = parsedCABundle
+
+	return cg, nil
+}
+
+func (cg *ClientCertificateGenerator) generate(r io.Reader, expiration time.Time, userMeta dbplugin.UsernameMetadata) (*certutil.CertBundle, string, error) {
+	commonName, err := cg.cnProducer.Generate(userMeta)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Set defaults
+	keyBits := cg.KeyBits
+	signatureBits := cg.SignatureBits
+	switch cg.KeyType {
+	case "rsa":
+		if keyBits == 0 {
+			keyBits = 2048
+		}
+		if signatureBits == 0 {
+			signatureBits = 256
+		}
+	case "ec":
+		if keyBits == 0 {
+			keyBits = 256
+		}
+		if signatureBits == 0 {
+			signatureBits = keyBits
+		}
+	case "ed25519":
+		// key_bits ignored
+		if signatureBits == 0 {
+			signatureBits = 256
+		}
+	}
+
+	subject := pkix.Name{
+		CommonName: commonName,
+		// Additional subject DN options intentionally omitted for now
```

Comment:
We should find out if this is okay. Perhaps cryptosec can help here!

Reply:
What database is this for? I think the short answer is it depends on the auth model of the database itself: does it limit permissions based on other subject attributes?

Reply:
@cipherboy This is for MongoDB Atlas secrets engine. Here is the PR that's on the db side: https://github.com/hashicorp/vault-plugin-database-mongodbatlas/pull/52

Reply:
Right, but how is MongoDB Atlas _doing_ the authentication? What attributes are there that it is validating against? Are there custom extension that are important? Or is it just "any certificate signed by a given CA is accepted"? &c.

Happy to talk on a call if you'd like to discuss it more. :-)

Reply:
I can explain some of the details around how MongoDB Atlas is doing the authentication. Might be easier over a quick call if you don't mind. I'll set something up!

Reply:
This should definitely be ok from what I saw. Just note that `OU/O/DN` might need to be set, depending on what the `tlsListener` had set on the MongoDB side :-)